### ISSUE TITLE
[Suggestion] modular language system

### ISSUE THREAD

**Issue by Jan200101** [Cog-Creators/Red-DiscordBot#706]
a system where red chooses text from a json file intergrated in 
`/data/red/languages/<language prefix>/lang.json`
there red would fetch strings to print.

It would allow for better usage on servers based on other languages and for custom language packs that include e.g. emojis

EDIT:
Translations are getting added and translations can already be submitted at [Crowdin](https://crowdin.com/project/red-discordbot)

**Comment by tekulvw** [comment]
It's called i18n, we've got methods of implementing it but never agreed on where the responsibility for generating translations should lie

**Comment by Jan200101** [comment]
translation could be made from the community.
Many other services do the same but I do see how that could go wrong.

Going with google translate would be the worst choice by far and some language like german or dutch could be translated by already existing members in the discord

**Comment by palmtree5** [comment]
There's too many problems with this. Sure, it'd be great if it were just core red but you've got all the third party cogs that would need translated. Change a string? Time to re-translate

**Comment by Jan200101** [comment]
It was just a suggestion for core.
3th party cogs would still need a translation

**Comment by palmtree5** [comment]
That's all well and good but something would probably have to be in core to load in the translations from 3rd party cogs

**Comment by Jan200101** [comment]
the 3th party cogs would do that themself a method for fetching language files would be nice though.

**Comment by palmtree5** [comment]
So your suggestion regarding third-party cogs is for every cog to have completely redundant code in them for handling translations?

**Comment by Jan200101** [comment]
my suggestion is making a language module base every cog can use to get strings out of language files designed for the cog.

**Comment by palmtree5** [comment]
The main issue though is that translation (whether for core or for third-party cogs) would have to be a community project. Can't really get away with putting core on the community while putting third-party stuff on the creators themselves. Not everyone knows a wide variety of languages

**Comment by Jan200101** [comment]
well thats true... havent really thought about how the cog creator would get the translation.

Another way of actually working through the problem is fetching the string it spits out translating it (if it doenst have dedicated translation file) and giving it to the user

**Comment by palmtree5** [comment]
Things that I can immediately think of that would need to be translated:

- docstrings
- command outputs

The entire thing would need to be per-server

**Comment by Conobi** [comment]
Good idea, it would be amazing.

**Comment by tekulvw** [comment]
Minus the troubles of actually getting translations, Thane just brought to my attention of actually doing the translating of docstrings at run time. We can put hooks into a custom HelpFormatter to do the heavy lifting for us.

**Comment by Kowlin** [comment]
Getting the translations is as easy as using a i18n community platform, we can reward them a bit with a role in the server or something related. Community outsourcing would be great for this.

**Comment by Twentysix26** [comment]
I've said this on Discord but I'll put my thoughts here too  

My main concerns about this are not technical, I'm sure there are ways to accomplish this in a mostly clean way, but rather logistics.
I'd say that the main strength of this bot is 3rd party content. The overwhelming majority of 3rd party cogs repositories have a very small workforce, with only the owner working on it and the occasional contribution.
Good quality translations are not something that can be done with automated tools.

I just don't picture most 3rd party cog devs wanting to bother with multiple translations too, which of course they'll have to keep updated as the time passes and content is added / edited.
Which in turn means that whoever will be interested in translations, will find either no translation at all for a lot of cogs or even an outdated one (picture sentences here and there defaulting to english, due to the cog being up to date but the translation still isn't).  

To recap, while this sounds neat in theory, translations need to be made and maintained from people who know the language, something that might be more effort than worth for 3rd party cog developers considering the very small number that would benefit from them.  

And for this very small number (most of our userbase is english speaking and the requests for a translated bot have been very few since I started this) translations are something I've never been particularly interested in implementing.

**Comment by tekulvw** [comment]
My only counter to this is that partial translations, however sparse and far between they are, are still better than none at all.

You're right in that a large majority of 3rd party cogs will never see translations and I think that's okay because the people requesting translations would be more than willing to help out on the cogs that they use and that's what matters. We don't need 100% coverage and they don't need to be perfect, we just need a mechanism to allow them.

**Comment by palmtree5** [comment]
https://crowdin.com/project/red-discordbot

**Comment by tekulvw** [comment]
Present in V3.
